FirebaseMLModelDownloader/Sources/ModelDownloadTask.swift

```diff
@@ -94,93 +81,96 @@ class ModelDownloadTask: NSObject {
-  /// Asynchronously download model file to device.
-  func resumeModelDownload() {
-    guard downloadStatus == .notStarted else { return }
-    let downloadTask = downloadSession.downloadTask(with: remoteModelInfo.downloadURL)
-    downloadTask.resume()
-    downloadStatus = .inProgress
-    self.downloadTask = downloadTask
-  }
-}
-
-/// Extension to handle delegate methods.
-extension ModelDownloadTask: URLSessionDownloadDelegate {
   /// Name for model file stored on device.
   var downloadedModelFileName: String {
     return "fbml_model__\(appName)__\(remoteModelInfo.name)"
   }
 
-  func urlSession(_ session: URLSession, taskIsWaitingForConnectivity task: URLSessionTask) {
-    // TODO: Handle waiting for connectivity, if needed.
-  }
-
-  func urlSession(_ session: URLSession, didBecomeInvalidWithError error: Error?) {
-    // TODO: Handle configuration errors.
-  }
-
-  /// Handle client-side errors.
-  func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
-    // TODO: Log this.
-    guard task == downloadTask else { return }
-    guard let error = error else { return }
-    /// Unable to resolve hostname or connect to host.
-    DispatchQueue.main.async {
-      self.downloadHandlers
-        .completion(.failure(.internalError(description: error.localizedDescription)))
+  func download() {
+    downloader.downloadFile(with: remoteModelInfo.downloadURL,
+                            progressHandler: { downloadedBytes, totalBytes in
+                              /// Fraction of model file downloaded.
+                              let calculatedProgress = Float(downloadedBytes) / Float(totalBytes)
+                              DispatchQueue.main.async {
+                                self.downloadHandlers.progressHandler?(calculatedProgress)
+                              }
+                            },
+                            configurationErrorHandler: { error in
+                              /// Unable to configure download session.
+                              DispatchQueue.main.async {
+                                self.downloadHandlers.completion(.failure(.failedPrecondition))
+                              }
+                            },
+                            downloadErrorHandler: { error in
+                              /// Unable to resolve hostname or connect to host.
+                              DispatchQueue.main.async {
+                                self.downloadHandlers
+                                  .completion(.failure(.internalError(description: ModelDownloadTask
+                                      .ErrorDescription
+                                      .invalidHostName(error
+                                        .localizedDescription))))
+                              }
+                            }) { response, location in
+      /// Model download completed.
+      guard let response = response as? HTTPURLResponse else {
```

Comment:
nit: This validation may be moved to `FileDownloader`

Reply:
Same as above. What if we'd like to pass a different error to the final completion based on whether it's a config error, or a HTTP response error, or something else? Or do we want the download error to be fairly opaque and logging to be more specific about the error?

Reply:
I think the approach described in the [comment](https://github.com/firebase/firebase-ios-sdk/pull/7352#discussion_r562863883) should be relevant here as well - you may pass a specific error in this case to be handled by ModelDownloadTask as needed.